candig_federation/operations.py

```diff
@@ -113,6 +117,90 @@ def delete_server(server_id):
     return result, 200
 
 
+def list_external_services():
+    """
+    :return: Dictionary of registered external services.
+    """
+    services = get_registered_external_services()
+    if services is not None:
+        result = map(lambda x: x["service"], services.values())
+        return list(result), 200
+    logger.debug(f"Couldn't list services", connexion.request)
+    return {"message": "Couldn't list services"}, 500
+
+
+async def add_external_service():
+    """
+    :return: Service added.
+    """
+    if not is_site_admin(connexion.request):
+        return {"message": "User is not authorized to POST"}, 403
+    try:
+        req = await connexion.request.json()
+        if req is not None and 'service' in req:
+            new_service = register_external_service(req)
+            if new_service is None:
+                return {"message": f"Service matching {new_service['service']} already present"}, 200
+
+            # the new service user needs to be CanDIG-authorized:
+            headers = {
+                "Content-Type": "application/json; charset=utf-8"
+            }
+
+            # first, preapprove the service user:
+            headers["Authorization"] = f"Bearer {get_auth_token(connexion.request)}"
+            response = requests.post(
+                f"{CANDIG_INGEST_PUBLIC_URL}/user/preapproved/{new_service["user"]}",
+                headers=headers
+            )
+
+            # then, request authorization for the service user
+            headers["Authorization"] = f"Bearer {new_service['token']}"
+            response = requests.post(
+                f"{CANDIG_INGEST_PUBLIC_URL}/user/pending/request",
+                headers=headers
+            )
+
+            if response.status_code != 200:
+                return {"message": f"Service user {new_service["user"]} could not be authorized: {response.text}"}, response.status_code
```

Comment:
In the lines before this, the service user is made to request authorization, but we don't actually try to approve the auth request yet -- is that meant to be a separate step?